app/posthandler/refund_gas_remaining.go

```diff
@@ -0,0 +1,144 @@
+package posthandler
+
+import (
+	"fmt"
+
+	"cosmossdk.io/errors"
+	"cosmossdk.io/math"
+	"github.com/celestiaorg/celestia-app/pkg/appconsts"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	authkeeper "github.com/cosmos/cosmos-sdk/x/auth/keeper"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	feegrantkeeper "github.com/cosmos/cosmos-sdk/x/feegrant/keeper"
+)
+
+// MaxPortionOfFeeToRefund is the maximum portion of the fee that can be refunded.
+var MaxPortionOfFeeToRefund = sdk.NewDecWithPrec(5, 1) // 50%
+
+// RefundGasCost is the amount of gas consumed during the execution of this
+// posthandler. If a tx reaches this posthandler with gas remaining in excess of
+// this amount, then a refund will be issued for the gas remaining -
+// RefundGasCost.
+//
+// NOTE: this value was determined empirically.
+const RefundGasCost = 15_000
+
+// RefundGasRemainingDecorator handles refunding a portion of the tx fee that was
+// originally deducted from the feepayer but was not needed because the tx
+// consumed less gas than the gas limit.
+type RefundGasRemainingDecorator struct {
+	accountKeeper  authkeeper.AccountKeeper
+	bankKeeper     authtypes.BankKeeper
+	feegrantKeeper feegrantkeeper.Keeper
+}
+
+// NewRefundGasRemainingDecorator returns a new RefundGasRemainingDecorator.
+func NewRefundGasRemainingDecorator(ak authkeeper.AccountKeeper, bk authtypes.BankKeeper, fk feegrantkeeper.Keeper) RefundGasRemainingDecorator {
+	return RefundGasRemainingDecorator{
+		accountKeeper:  ak,
+		bankKeeper:     bk,
+		feegrantKeeper: fk,
+	}
+}
+
+// AnteHandle implements the cosmos-sdk AnteHandler interface. Note: the
+// AnteHandler interface is also used for post-handlers.
+func (d RefundGasRemainingDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	if err := d.maybeRefund(ctx, tx, simulate); err != nil {
+		return ctx, err
+	}
+	return next(ctx, tx, simulate)
+}
```

Comment:
blocking

can we get an if statement for the app version here along with a quick unit test to make sure its not making any changes for appversion v1?

Reply:
Great call! d69402889f896b2dc2d548860c5dce2f232d8b56